File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/claim_ownership/ClaimOwnershipWizardViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions", "LongMethod")

package app.pantopus.android.ui.screens.homes.claim_ownership

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.analytics.Analytics
import app.pantopus.android.data.analytics.AnalyticsEvent
import app.pantopus.android.data.analytics.AnalyticsResult
import app.pantopus.android.data.api.models.homes.SubmitClaimRequest
import app.pantopus.android.data.api.models.homes.UploadEvidenceRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.data.network.NetworkMonitor
import app.pantopus.android.ui.screens.shared.wizard.WizardChrome
import app.pantopus.android.ui.screens.shared.wizard.WizardLeadingControl
import app.pantopus.android.ui.screens.shared.wizard.WizardModel
import app.pantopus.android.ui.screens.shared.wizard.WizardProgressLabel
import app.pantopus.android.ui.screens.shared.wizard.WizardSecondaryCta
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg key for the home being claimed. */
const val CLAIM_OWNERSHIP_HOME_ID_KEY: String = "homeId"

/**
 * Aggregate UI state for the Claim Ownership wizard. Combined so the
 * screen derives the [WizardChrome] from a single state read.
 */
data class ClaimOwnershipUiState(
    val currentStep: ClaimOwnershipStep = ClaimOwnershipStep.Start,
    val slots: Map<ClaimEvidenceSlot, ClaimSlotState> =
        ClaimEvidenceSlot.entries.associateWith { ClaimSlotState.Empty },
    val note: String = "",
    val isSubmitting: Boolean = false,
    val submitError: String? = null,
) {
    val bothSlotsHaveFiles: Boolean
        get() = slots.values.all { it.hasFile }

    val anySlotHasFile: Boolean
        get() = slots.values.any { it.hasFile }
}

/**
 * Drives the 3-step claim-ownership wizard. Calls:
 *  1. `POST /api/homes/:id/ownership-claims` to create the claim
 *  2. For each evidence file:
 *      a. `POST /api/files/upload` (multipart) → URL
 *      b. `POST /api/homes/:id/ownership-claims/:claimId/evidence`
 *         with `storage_ref = <url>`
 *
 * Backend deviations flagged in the PR description:
 *  - `submitClaimSchema` does NOT accept a `note` field; the wizard's
 *    optional textarea is piped into evidence metadata on the first
 *    file.
 *  - The evidence endpoint takes JSON `storage_ref`, not multipart —
 *    we route bytes through `/api/files/upload` first.
 */
@HiltViewModel
open class ClaimOwnershipWizardViewModel
    @Inject
    constructor(
        private val repository: HomesRepository,
        private val networkMonitor: NetworkMonitor,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel(),
        WizardModel {
        private val homeId: String =
            requireNotNull(savedStateHandle[CLAIM_OWNERSHIP_HOME_ID_KEY]) {
                "ClaimOwnershipWizardViewModel requires a '$CLAIM_OWNERSHIP_HOME_ID_KEY' nav arg."
            }

        private val _state = MutableStateFlow(ClaimOwnershipUiState())
        val state: StateFlow<ClaimOwnershipUiState> = _state.asStateFlow()

        /** One-shot navigation events the screen reacts to. */
        val pendingEvent = MutableStateFlow<ClaimOwnershipOutboundEvent?>(null)

        // MARK: - WizardModel

        override val chrome: WizardChrome
            get() = computeChrome(_state.value)

        override fun onLeading() {
            when (_state.value.currentStep) {
                ClaimOwnershipStep.Start -> pendingEvent.value = ClaimOwnershipOutboundEvent.Dismiss
                ClaimOwnershipStep.Upload -> transitionTo(ClaimOwnershipStep.Start)
                ClaimOwnershipStep.Success -> pendingEvent.value = ClaimOwnershipOutboundEvent.Dismiss
            }
        }

        override fun onDiscard() {
            pendingEvent.value = ClaimOwnershipOutboundEvent.Dismiss
        }

        override fun onPrimary() {
            when (_state.value.currentStep) {
                ClaimOwnershipStep.Start -> transitionTo(ClaimOwnershipStep.Upload)
                ClaimOwnershipStep.Upload -> viewModelScope.launch { submit() }
                ClaimOwnershipStep.Success ->
                    pendingEvent.value = ClaimOwnershipOutboundEvent.OpenClaimsList
            }
        }

        override fun onSecondary() {
            if (_state.value.currentStep == ClaimOwnershipStep.Success) {
                pendingEvent.value = ClaimOwnershipOutboundEvent.Dismiss
            }
        }

        // MARK: - Slot management

        fun picked(slot: ClaimEvidenceSlot, file: ClaimPickedFile) {
            _state.update { current ->
                current.copy(
                    slots = current.slots.toMutableMap().apply { put(slot, ClaimSlotState.Picked(file)) },
                    submitError = null,
                )
            }
        }

        fun remove(slot: ClaimEvidenceSlot) {
            _state.update { current ->
                current.copy(slots = current.slots.toMutableMap().apply { put(slot, ClaimSlotState.Empty) })
            }
        }

        fun setNote(value: String) {
            _state.update { it.copy(note = value) }
        }

        fun acknowledgeEvent() {
            pendingEvent.value = null
        }

        // MARK: - Submit

        private suspend fun submit() {
            val current = _state.value
            if (!current.bothSlotsHaveFiles || current.isSubmitting) return
            if (!networkMonitor.isOnline.value) {
                _state.update {
                    it.copy(
                        submitError = "You're offline. Try again when you're back online.",
                    )
                }
                return
            }
            _state.update { it.copy(isSubmitting = true, submitError = null) }

            val claimResult = repository.submitClaim(homeId, SubmitClaimRequest(method = "doc_upload"))
            val claim =
                when (claimResult) {
                    is NetworkResult.Success -> claimResult.data.claim
                    is NetworkResult.Failure -> {
                        Analytics.track(AnalyticsEvent.CtaClaimOwnershipSubmit(AnalyticsResult.ERROR))
                        _state.update {
                            it.copy(isSubmitting = false, submitError = "Couldn't submit. Retry.")
                        }
                        return
                    }
                }
            val claimId =
                claim.id ?: run {
                    Analytics.track(AnalyticsEvent.CtaClaimOwnershipSubmit(AnalyticsResult.ERROR))
                    _state.update {
                        it.copy(
                            isSubmitting = false,
                            submitError = "We're already working on a claim for this home.",
                        )
                    }
                    return
                }

            // Upload each slot's bytes, then register the URL as evidence.
            for ((index, slot) in ClaimEvidenceSlot.entries.withIndex()) {
                val file = current.slots[slot]?.pickedFile ?: continue
                markSlot(slot, ClaimSlotState.Uploading(file, 0.4f))
                val uploadResult = repository.uploadFile(file.filename, file.mimeType, file.bytes)
                val fileUrl =
                    when (uploadResult) {
                        is NetworkResult.Success -> uploadResult.data.file.url
                        is NetworkResult.Failure -> {
                            markSlot(slot, ClaimSlotState.Failed(file, "Upload failed"))
                            failSubmit()
                            return
                        }
                    }
                markSlot(slot, ClaimSlotState.Uploading(file, 0.8f))
                val metadata: Map<String, String>? =
                    if (index == 0 && current.note.trim().isNotEmpty()) {
                        mapOf("note" to current.note.trim())
                    } else {
                        null
                    }
                val evidenceResult =
                    repository.uploadEvidence(
                        homeId = homeId,
                        claimId = claimId,
                        request =
                            UploadEvidenceRequest(
                                evidenceType = slot.backendType,
                                storageRef = fileUrl,
                                metadata = metadata,
                            ),
                    )
                when (evidenceResult) {
                    is NetworkResult.Success -> markSlot(slot, ClaimSlotState.Uploaded(file, fileUrl))
                    is NetworkResult.Failure -> {
                        markSlot(slot, ClaimSlotState.Failed(file, "Upload failed"))
                        failSubmit()
                        return
                    }
                }
            }

            Analytics.track(AnalyticsEvent.CtaClaimOwnershipSubmit(AnalyticsResult.SUCCESS))
            _state.update { it.copy(isSubmitting = false) }
            transitionTo(ClaimOwnershipStep.Success)
        }

        private fun failSubmit() {
            _state.update {
                it.copy(
                    isSubmitting = false,
                    submitError = "Couldn't submit. Retry.",
                )
            }
        }

        private fun markSlot(slot: ClaimEvidenceSlot, value: ClaimSlotState) {
            _state.update { current ->
                current.copy(slots = current.slots.toMutableMap().apply { put(slot, value) })
            }
        }

        // MARK: - Step transitions

        private fun transitionTo(step: ClaimOwnershipStep) {
            _state.update { it.copy(currentStep = step) }
            Analytics.track(AnalyticsEvent.ScreenClaimOwnershipStepViewed(step.name))
        }

        // MARK: - Chrome derivation

        private fun computeChrome(state: ClaimOwnershipUiState): WizardChrome =
            when (state.currentStep) {
                ClaimOwnershipStep.Start ->
                    WizardChrome(
                        title = "Claim ownership",
                        progressLabel = WizardProgressLabel.StepOf(1, 3),
                        progressFraction = 1f / 3f,
                        leading = WizardLeadingControl.Close,
                        primaryCtaLabel = "Start claim",
                        primaryCtaEnabled = true,
                        secondaryCta = null,
                        isSubmitting = false,
                        dirty = false,
                        showsProgressBar = true,
                    )
                ClaimOwnershipStep.Upload ->
                    WizardChrome(
                        title = "Claim ownership",
                        progressLabel = WizardProgressLabel.StepOf(2, 3),
                        progressFraction = 2f / 3f,
                        leading = WizardLeadingControl.Back,
                        primaryCtaLabel = "Submit claim",
                        primaryCtaEnabled = state.bothSlotsHaveFiles && !state.isSubmitting,
                        secondaryCta = null,
                        isSubmitting = state.isSubmitting,
                        dirty = state.anySlotHasFile || state.note.isNotEmpty(),
                        showsProgressBar = true,
                    )
                ClaimOwnershipStep.Success ->
                    WizardChrome(
                        title = "Claim ownership",
                        progressLabel = WizardProgressLabel.Hidden,
                        progressFraction = null,
                        leading = WizardLeadingControl.Close,
                        primaryCtaLabel = "View status",
                        primaryCtaEnabled = true,
                        secondaryCta = WizardSecondaryCta("Back to home", testTag = "claimOwnershipBackToHome"),
                        isSubmitting = false,
                        dirty = false,
                        showsProgressBar = false,
                    )
            }
    }

File: frontend/apps/ios/Pantopus/Features/Homes/ClaimOwnership/ClaimOwnershipWizardViewModel.swift
//
//  ClaimOwnershipWizardViewModel.swift
//  Pantopus
//
//  3-step claim-ownership wizard. Backend flow:
//   1. POST /api/homes/:id/ownership-claims  (claim_type=owner, method=doc_upload)
//   2. For each evidence file:
//        a. POST /api/files/upload (multipart) → file URL
//        b. POST /api/homes/:id/ownership-claims/:claimId/evidence with storage_ref
//   3. On all-success: advance to .success
//   4. On any failure: stay on .upload, mark the failing slot, preserve files
//
//  Backend reality vs P20 spec — flagged in the PR description:
//  - submitClaimSchema does NOT accept a `note` field. The textarea
//    value is sent as `metadata.note` on the FIRST evidence upload.
//  - The evidence endpoint takes JSON `storage_ref`, not multipart.
//    Real bytes go through `/api/files/upload` first.
//
// swiftlint:disable function_body_length

import Foundation
import Logging
import Observation

/// Outbound events the wizard view must react to.
public enum ClaimOwnershipOutboundEvent: Sendable, Equatable {
    case dismiss
    case openClaimsList
}

/// ViewModel backing `ClaimOwnershipWizardView`.
@Observable
@MainActor
final class ClaimOwnershipWizardViewModel: WizardModel {
    // MARK: - Published state

    private(set) var currentStep: ClaimOwnershipStep = .start
    var slots: [ClaimEvidenceSlot: ClaimSlotUiState] = [:]
    var note: String = ""
    private(set) var startContent: ClaimOwnershipStartContent
    private(set) var isSubmitting: Bool = false
    private(set) var submitError: String?
    var pendingEvent: ClaimOwnershipOutboundEvent?

    /// Server-side claim id once `POST /ownership-claims` succeeds. Held
    /// across retry attempts so a partial-success → retry doesn't create
    /// a duplicate claim row server-side.
    private var pendingClaimId: String?
    /// File URLs successfully pushed through `/api/files/upload` whose
    /// evidence registration later failed. Held so retry can POST the
    /// evidence call directly with the existing `storage_ref` instead
    /// of re-uploading the bytes (which would orphan the prior file).
    private var pendingUploadURLs: [ClaimEvidenceSlot: String] = [:]

    // MARK: - Init

    private let homeId: String
    private let api: APIClient
    private let uploader: MultipartUploader
    private let logger = Logger(label: "app.pantopus.ios.ClaimOwnershipWizard")

    init(
        homeId: String,
        api: APIClient = .shared,
        uploader: MultipartUploader = .shared,
        startContent: ClaimOwnershipStartContent? = nil
    ) {
        self.homeId = homeId
        self.api = api
        self.uploader = uploader
        self.startContent = startContent ?? ClaimOwnershipSampleData.startContent(for: homeId)
        for slot in ClaimEvidenceSlot.allCases {
            slots[slot] = .empty
        }
    }

    // MARK: - WizardModel

    var chrome: WizardChrome {
        switch currentStep {
        case .start:
            WizardChrome(
                title: "Claim ownership",
                progressLabel: .stepOf(current: 1, total: 3),
                progressFraction: 1.0 / 3.0,
                leading: .close,
                primaryCTALabel: "Start claim",
                primaryCTAEnabled: true,
                secondaryCTA: nil,
                isSubmitting: false,
                // Once the user has filled any slot or typed a note on the
                // upload step, going back to Start must still surface the
                // discard-confirm so an X tap doesn't dump the in-memory
                // bytes silently.
                dirty: anySlotHasFile || !note.isEmpty,
                showsProgressBar: true
            )
        case .upload:
            WizardChrome(
                title: "Claim ownership",
                progressLabel: .stepOf(current: 2, total: 3),
                progressFraction: 2.0 / 3.0,
                leading: .back,
                primaryCTALabel: "Submit claim",
                primaryCTAEnabled: bothSlotsHaveFiles && !isSubmitting,
                secondaryCTA: nil,
                isSubmitting: isSubmitting,
                dirty: anySlotHasFile || !note.isEmpty,
                showsProgressBar: true
            )
        case .success:
            WizardChrome(
                title: "Claim ownership",
                progressLabel: .hidden,
                progressFraction: nil,
                leading: .close,
                primaryCTALabel: "View status",
                primaryCTAEnabled: true,
                secondaryCTA: WizardSecondaryCTA(label: "Back to home", identifier: "claimOwnership_backToHome"),
                isSubmitting: false,
                dirty: false,
                showsProgressBar: false
            )
        }
    }

    func leadingTapped() {
        switch currentStep {
        case .start:
            pendingEvent = .dismiss
        case .upload:
            currentStep = .start
        case .success:
            pendingEvent = .dismiss
        }
    }

    func discardConfirmed() {
        pendingEvent = .dismiss
    }

    func primaryTapped() {
        switch currentStep {
        case .start:
            currentStep = .upload
        case .upload:
            Task { await submit() }
        case .success:
            pendingEvent = .openClaimsList
        }
    }

    func secondaryTapped() {
        // Only fires on success — "Back to home".
        pendingEvent = .dismiss
    }

    // MARK: - Slot management

    func picked(_ slot: ClaimEvidenceSlot, file: ClaimPickedFile) {
        slots[slot] = .picked(file: file)
        // Picking a new file invalidates any prior URL we'd cached for
        // this slot — the next submit must re-upload these bytes.
        pendingUploadURLs[slot] = nil
        submitError = nil
    }

    func remove(_ slot: ClaimEvidenceSlot) {
        slots[slot] = .empty
        pendingUploadURLs[slot] = nil
    }

    /// Surface a "file too large" error inline rather than letting the
    /// upload round-trip to a 413. Called by the picker when the user
    /// selects a file over `CLAIM_FILE_MAX_BYTES`.
    func fileTooLarge(for _: ClaimEvidenceSlot) {
        submitError = "That file is over 10 MB. Try a smaller photo."
    }

    var bothSlotsHaveFiles: Bool {
        slots.values.allSatisfy(\.hasFile)
    }

    var anySlotHasFile: Bool {
        slots.values.contains { $0.hasFile }
    }

    // MARK: - Submit

    func submit() async {
        guard bothSlotsHaveFiles, !isSubmitting else { return }
        if !NetworkMonitor.shared.isOnline {
            submitError = "You're offline. Try again when you're back online."
            return
        }
        isSubmitting = true
        submitError = nil
        defer { isSubmitting = false }

        // Step 1: create the claim — but only once across retry attempts.
        // Holding the id in `pendingClaimId` keeps a partial-success retry
        // from creating a duplicate claim row server-side.
        let claimId: String
        if let existing = pendingClaimId {
            claimId = existing
        } else {
            let claimResponse: SubmitClaimResponse
            do {
                claimResponse = try await api.request(
                    HomesEndpoints.submitClaim(
                        homeId: homeId,
                        request: SubmitClaimRequest(method: "doc_upload")
                    )
                )
            } catch {
                submitError = "Couldn't submit. Retry."
                logger.warning("Claim submit failed: \(error)")
                Analytics.track(.ctaClaimOwnershipSubmit(result: .error))
                return
            }
            guard let id = claimResponse.claim.id else {
                // Opaque-handshake path can return nil claim id when a
                // duplicate exists. Surface a friendly message rather
                // than failing silently.
                submitError = "We're already working on a claim for this home."
                Analytics.track(.ctaClaimOwnershipSubmit(result: .error))
                return
            }
            claimId = id
            pendingClaimId = id
        }

        // Step 2: upload each slot's file then register evidence. Skip
        // any slot we already finished on a prior attempt, and skip the
        // upload step for slots whose bytes are already in storage —
        // both retry-paths exist so partial failures don't repeat work.
        for (index, slot) in ClaimEvidenceSlot.allCases.enumerated() {
            if case .uploaded = slots[slot] { continue }
            guard let file = slots[slot]?.pickedFile else { continue }
            let metadata: [String: String]? =
                index == 0 && !note.trimmingCharacters(in: .whitespaces).isEmpty
                    ? ["note": note] : nil
            do {
                let fileURL: String
                if let cached = pendingUploadURLs[slot] {
                    fileURL = cached
                } else {
                    slots[slot] = .uploading(file: file, fraction: 0.4)
                    let upload = try await uploader.uploadFile(
                        MultipartFile(
                            fieldName: "file",
                            filename: file.filename,
                            mimeType: file.mimeType,
                            data: file.data
                        ),
                        formFields: ["file_type": "claim_evidence", "visibility": "private"]
                    )
                    fileURL = upload.file.url
                    pendingUploadURLs[slot] = fileURL
                }
                slots[slot] = .uploading(file: file, fraction: 0.8)
                _ = try await api.request(
                    HomesEndpoints.uploadEvidence(
                        homeId: homeId,
                        claimId: claimId,
                        request: UploadEvidenceRequest(
                            evidenceType: slot.backendType,
                            storageRef: fileURL,
                            metadata: metadata
                        )
                    )
                ) as UploadEvidenceResponse
                slots[slot] = .uploaded(file: file, fileURL: fileURL)
                // Evidence row exists — no need to keep the URL cache.
                pendingUploadURLs[slot] = nil
            } catch {
                logger.warning("Evidence upload failed for slot \(slot.rawValue): \(error)")
                slots[slot] = .failed(file: file, message: "Upload failed")
                submitError = "Couldn't submit. Retry."
                Analytics.track(.ctaClaimOwnershipSubmit(result: .error))
                return
            }
        }

        // All uploads succeeded — advance to success.
        Analytics.track(.ctaClaimOwnershipSubmit(result: .success))
        currentStep = .success
    }

    func acknowledgePendingEvent() {
        pendingEvent = nil
    }
}
